File: gateway/adapters/telegram.py
from __future__ import annotations

import logging
from typing import Callable

from .base import ChannelAdapter

log = logging.getLogger(__name__)


class TelegramAdapter(ChannelAdapter):
    """Telegram adapter using python-telegram-bot (polling mode)."""

    def __init__(self, token: str):
        self._token = token
        self._app = None

    async def start(self, on_message: Callable) -> None:
        from telegram.ext import Application, MessageHandler, filters

        app = Application.builder().token(self._token).build()
        self._app = app

        async def handler(update, context):
            msg = update.message
            if not msg or not msg.text:
                return
            peer_id = str(msg.chat_id)

            async def deliver(pid: str, text: str) -> None:
                await context.bot.send_message(chat_id=int(pid), text=text)

            await on_message("telegram", peer_id, msg.text, deliver)

        app.add_handler(MessageHandler(filters.TEXT & ~filters.COMMAND, handler))
        log.info("Telegram adapter starting (polling)")
        await app.initialize()
        await app.start()
        await app.updater.start_polling()
        # keep running until cancelled
        try:
            import asyncio
            await asyncio.Event().wait()
        finally:
            await app.updater.stop()
            await app.stop()
            await app.shutdown()

    async def send(self, peer_id: str, text: str) -> None:
        if self._app:
            await self._app.bot.send_message(chat_id=int(peer_id), text=text)

File: src/gateway/adapters/telegram.ts
import { ChannelAdapter } from "./base.ts";
import type { OnMessageFn } from "../router.ts";

export class TelegramAdapter extends ChannelAdapter {
  private bot: unknown = null;

  constructor(private readonly token: string) {
    super();
  }

  async start(onMessage: OnMessageFn): Promise<void> {
    const { Bot } = await import("grammy");

    const bot = new (Bot as new (token: string) => {
      on(event: string, handler: (ctx: Record<string, unknown>) => Promise<void>): void;
      start(): Promise<void>;
      api: { sendMessage(chatId: number, text: string): Promise<unknown> };
    })(this.token);

    this.bot = bot;

    bot.on("message:text", async (ctx: Record<string, unknown>) => {
      const msg = ctx["message"] as Record<string, unknown> | undefined;
      if (!msg) return;
      const text = msg["text"] as string | undefined;
      if (!text) return;
      const chatId = msg["chat_id"] as number | undefined ?? (msg["chat"] as Record<string, unknown>)?.["id"] as number;
      const peerId = String(chatId);

      await onMessage("telegram", peerId, text, async (_pid: string, chunk: string) => {
        await bot.api.sendMessage(chatId, chunk);
      });
    });

    console.log("Telegram adapter starting (polling)");
    await bot.start();
  }

  async send(peerId: string, text: string): Promise<void> {
    if (this.bot) {
      const b = this.bot as { api: { sendMessage(chatId: number, text: string): Promise<unknown> } };
      await b.api.sendMessage(parseInt(peerId, 10), text);
    }
  }
}
